core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -166,6 +195,72 @@ class CloudSqlInstance {
     }
   }
 
+  private static CredentialFactory loadCredentialFactory() {
+    String userCredentialFactoryClassName = System.getProperty(
+        CredentialFactory.CREDENTIAL_FACTORY_PROPERTY);
+
+    CredentialFactory credentialFactory;
+    if (userCredentialFactoryClassName != null) {
+      try {
+        credentialFactory =
+            (CredentialFactory)
+                Class.forName(userCredentialFactoryClassName).newInstance();
+      } catch (Exception err) {
+        throw new RuntimeException(err);
+      }
+    } else {
+      credentialFactory = new ApplicationDefaultCredentialFactory();
+    }
+
+    return credentialFactory;
+  }
+
+
+  private SQLAdmin createAdminApiClient() {
+    HttpRequestInitializer credential = credentialFactory.create();
+    HttpTransport httpTransport;
+    try {
+      httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    } catch (GeneralSecurityException | IOException err) {
+      throw new RuntimeException("Unable to initialize HTTP transport", err);
+    }
+
+    String rootUrl = System.getProperty(API_ROOT_URL_PROPERTY);
+    String servicePath = System.getProperty(API_SERVICE_PATH_PROPERTY);
+
+    JsonFactory jsonFactory = GsonFactory.getDefaultInstance();
+
+    SQLAdmin.Builder adminApiBuilder =
+        new SQLAdmin.Builder(httpTransport, jsonFactory, credential)
+            .setApplicationName(getUserAgents());
+    if (rootUrl != null) {
+      logTestPropertyWarning(API_ROOT_URL_PROPERTY);
+      adminApiBuilder.setRootUrl(rootUrl);
+    }
+    if (servicePath != null) {
+      logTestPropertyWarning(API_SERVICE_PATH_PROPERTY);
+      adminApiBuilder.setServicePath(servicePath);
+    }
+    return adminApiBuilder.build();
+  }
+
+
+  /**
+   * Returns the default string which is appended to the SQLAdmin API client User-Agent header.
+   */
+  static String getUserAgents() {
+    return String.join(" ", CoreSocketFactory.userAgents);
```

Comment:
I agree this could use some work. CloudSqlInstance instantiates the SqlAdmin api client, so it needs know the user agent. That's why I moved getUserAgent() here. I left CoreSocketFactory.userAgents in place to keep the change set small.